### ISSUE TITLE
Improve code comments and readability in Boolean.cpp

### ISSUE THREAD

**Issue by phkahler** [solvespace/solvespace#1357]
The code at the core of NURBS booleans is tricky to read and understand. In particular we mat want some(more) comments for the functions:

KeepRegion, KeepEdge, TagByClassifiedEdge

And maybe better names for the SShell::Class and SBspUv::Class or the variables that get those assignments.

SBspUv::Class is used to classify an edge as INSIDE, OUTSIDE, EDGE_PARALLEL, EDGE_ANTIPARALLEL to the trim polygon of a surface.

SShell::Class is similarly used to classify a region of a surface (bounded by an edge) as being INSIDE, OUTSIDE, COINC_SAME, COINC_OPP meaning inside or outside the other shell in a boolean, or coincident with some surface on the other shell while having the same or opposite normal as that other shell surface.

The duplicate names CLASS (for classification), and options INSIDE, OUTSIDE can be very confusing as it is harder to reason about the geometric meaning of these. Not sure if we should rename variables and or the enums CLASS, or even INSIDE, OUTSIDE to disambiguate them.

There are also 2 functions called ClassifyEdge. One for the Bsp/trim polygon and one in the SShell class. Some renaming here might help.

**Comment by phkahler** [comment]
@jwesthues  Regarding the merging of coincident surface. When performing a union or two disjoint shells that share a common surface, are all edges and surfaces kept during the boolean, and only merged afterward in MergeCoincidentSurfaces() ?

I'm looking at the edge and region classifier code. If two boxes of the same size are unioned at one face, should the original edges be kept or discarded?

Also in the boolean code, is [this line a bug?](https://github.com/solvespace/solvespace/blob/aa0d25e6ae7e718aa90365a39f4eccdecd62d42f/src/srf/boolean.cpp#L616)
My understanding is that the SShell::Class classification indicates if a region of a surface along an edge is inside, outside, or coincident with the shell. In this case it looks like these two are initialized as though they are the other classification of inside/outside the trim polygon of the surface containing the edge, but that's a different Class type that lives in the Bsp. We don't know if the surface is inside or outside the other shell at that point, and not even if it's inside or outside our own shell.

**Comment by jwesthues** [comment]
If the boundary of a shell contains e.g. two coplanar polygons, then we try to represent that as two loops in a single plane (represented as an `SSurface`), not one loop each in two coincident planes. The latter form is bad, e.g. because in a later Boolean operation, when we generate all the possible intersection curves (of surfaces from operand A against surfaces from operand B), we'll generate duplicates. The purpose of `MergeCoincidentSurfaces()` is to fix that.

That's a separate question from which edges to keep. In your example of the union between two identical faces, we need to keep one set of edges and discard the other. It's arbitrary which we keep. This kind of stuff is the reason why the union operation code isn't always symmetric in A and B, even though the operation is commutative.

I wouldn't completely trust my theory of which volumes, surfaces, and edges to keep (especially considering that we found a mistake in it when we added the intersection op). The highlighted line is at least not obviously a bug to me, though. I'd guess my theory was something like "this is an edge from the original shell, so we already know the indir is in and the outdir is out". I haven't studied long enough to say that confidently, though.

The 3d Boolean essentially requires an inner 2d Boolean for each surface. So the parallelism is deliberate, though the result is indeed pretty hard to follow. Perhaps we should call them all `UVEDGE_xxx` and `SURFACE_xxx` or something, so it's always clear what we're classifying.

**Comment by phkahler** [comment]
Maybe a shorter prefix for the surfaces. maybe I'll try these in my branch:
SURF_INSIDE
SURF_OUTSIDE
SURF_COINC_SAME
SURF_COINC_OPP

**Comment by phkahler** [comment]
I'm really stuck on SSurface::EdgeNormalsWithinSurface()  It makes no sense to  me why we would care about the edge normal pointing *out* of the trim polygon since that's not even part of the shell.

**Comment by jwesthues** [comment]
The basic idea is that we need to classify the volume and surface regions both inside and outside each trim curve edge (since we can't decide whether to keep or discard the edge without knowing both), so we need both the inner and outer normals. The comment notes that inner and outer normals aren't necessarily antiparallel when the surface is curved, which seems true, and I guess I wouldn't have added that complexity if that didn't solve a broken case. I'm afraid I don't remember much beyond that, though.

I developed all of these algorithms debugging graphically, with lots of temporary code to plot various edges and normals that never got checked in. I suspect these algorithms are near-impossible to understand or modify without that. It might be useful to recreate that in more permanent and polished form, some kind of (developer-only) interface to show all the edges and normals in a given Boolean, maybe even with lists in the text window to show and hide individual candidate edges (grouped by source, the original surface trim curves plus the intersections) and such.

**Comment by phkahler** [comment]
> The basic idea is that we need to classify the volume and surface regions both inside and outside each trim curve edge

And after much study and thinking about it, I understand how all of that can be inferred from what's going on at an edge. I have no idea how you arrived at this solution but it's really quite impressive. I'm toying with some ideas and may try rewriting parts of boolean.cpp. We know there are a few bugs but I'm mostly convinced that rewriting a few functions with careful consideration of all the cases will result in perfect booleans except maybe numerical issues.

There are 2 kinds of edge checks - existing edges, and those formed by intersections between shells. In each case up to 4 surface sections meet at the edge, Looking at a 2d cross section perpendicular to and through the middle of an edge should be all we need to determine which 2 surfaces need to be joined at that edge, and we can just discard the duplicates that aren't that one. The rules change based on the type of boolean of course.

I see a need for a high level strategy at least in regard to overlapping surfaces. In a union we need to discard the overlap region of either shell A or B, and that rule needs to be applied consistently iin all cases of edge classification. I think such a rule may be implicit already, so I'll try to figure that out so incremental rewrites to some of the tests can be done. Rewrites that will include consideration of curvature, which is needed to make #1291 work correctly (the curved surface will classify as coincident based on normals at the edge, but it actually goes inside the box).

BTW I think the reason edge normals (within a surface) may have been different on opposite sides is that you used to use a small delta (du,dv) to compute them? Still do? Using the current TangentsAt function they really should be exactly opposite directions.

**Comment by jwesthues** [comment]
I think the commit message from ae35b3595cf646be78ba6636baa2ad01a5cf118d may explain it, that they're indeed only different because it's a non-infinitesimal (du, dv), but that's deliberate in order to distinguish coincidences from tangencies.

A few papers describe polyhedron Booleans, and I think I got that basic idea of classifying based on regions around edges from there. There's almost nothing on curved surfaces though, and that brings considerable extra pain.

**Comment by phkahler** [comment]
@ruevs  After fixing some logic in KeepRegion I finally understand what "orig" is and "inOrig". The function is determining weather to keep a region of a surface based on the boolean operation and how that region relates to the other operand in the boolean. It seems like the only important thing is Inside/Outside/Coincident_same/Coincident_opposite and a flag for if we're talking about a region in opA or not (opB).

So turns out inOrig is being used as a flag to indicate if the region was inside the trim polygon of its own surface. That's why it can early exit this function. It's also odd to pass it all the way down here, and even more odd to use the SShell::Class enumeration to pass this flag around.

**Comment by ruevs** [comment]
@phkahler your understanding matches the understanding I developed while implementing intersection ;-)

**Comment by ruevs** [comment]
For cross reference ;-)
https://news.ycombinator.com/item?id=37634207
